testbed/testbed/otelcol_runner.go

```diff
@@ -0,0 +1,174 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package testbed
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/shirou/gopsutil/process"
+	"github.com/spf13/viper"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"go.opentelemetry.io/collector/config"
+	"go.opentelemetry.io/collector/config/configmodels"
+	"go.opentelemetry.io/collector/internal/version"
+	"go.opentelemetry.io/collector/service"
+)
+
+// OtelcolRunner defines the interface for configuring, starting and stopping one or more instances of
+// otelcol which will be the subject of testing being executed.
+type OtelcolRunner interface {
+	// PrepareConfig stores the provided YAML-based otelcol configuration file in the format needed by the otelcol
+	// instance(s) this runner manages. If successful, it returns the value for the --config command line parameter.
+	PrepareConfig(configStr string) (configFileName string, err error)
```

Comment:
From what I see PrepareConfig() is never called via a dynamic dispatch. If that is the case I do not see why this needs to be part of the interface, especially given that one of the implementations of OtelcolRunner, the InProcessCollector.PrepareConfig does not actually return any config file name. I think it will be nicer to delete this function declaration here and let each Runner have its own version of PrepareConfig() which does not necessarily have the same signature.
That is unless you are planning on having some generic code that requires dynamic dispatching of PrepareConfig because the code does not know which implementation of OtelcolRunner it deals with.

Reply:
I think the return values on OtelcolRunner's methods will be needed when interacting with remote instances. I was planning to address this when implementing the remote management implementation(s) and do any cleanup or design fixes at that point.
I have gone ahead and changed the PrepareConfig method to return a cleanup function which will never be nil but may do nothing. I went ahead and changed all the existing test cases to use the PrepareConfig method in a consistent manner including a defer using the cleanup method.

Reply:
Sounds good, thanks.